### DISCUSSION TITLE
Encryption

### DISCUSSION THREAD

**Discussion by odbuser2** [ArcadeData/arcadedb#535] [Q&A]
Will encryption be supported at various levels in ArcadeDB?

e.g.
By database for encryption at rest
By type
By property to protect specific data

**Comment by odbuser2** [comment]
Any thoughts on this?  In particular support for bcfips via the standard java.security.Provider could widen the use of arcadedb.

**Reply by lvca** [reply]
This is not a high priority, unless many people vote for it or we find a sponsor.

**Reply by tolgaulas** [reply]
My two cents is the core db should be as simple as possible, encryption can be handled by application level. If its a dire need, I think arcadedb will soon have user defined functions (https://github.com/ArcadeData/arcadedb/issues/700) that can be used for custom cryptography features by the implementors. (@lvca : 1 stone two birds :) )

**Reply by odbuser2** [reply]
I should have qualified the request with the need to have encryption at rest (db level).  Not encryption to satisfy every possible case (e.g. field level, rest, etc.).

Application handling of encryption is ok for "field" level encrypted but it will not address encryption at rest of the entire database.  Orientdb supported this between 2.2 and 3 (I thought it was also in 3+ but maybe not?). 

This shouldn't be hard as it could be a wrapper around the file io.  @lvca maybe you could point out the locations and any gotchas?  I can take a look.

Without "at rest" encryption via java.security, arcade can't be used in a huge chunk of the industry.

**Reply by lvca** [reply]
@odbuser2 We could support encryption at rest at the page level, so it would be efficient. The issue is with the size: the page has a fixed size, but with encryption it becomes variable. If it's shorter, we could fill it with blanks, but if larger we can't store it with the current architecture.

**Reply by lvca** [reply]
Silly question: what about using an encrypted folder to store the ArcadeDB database? In this way, the OS file system takes care of the encryption and it would be transparent for ArcadeDB.

**Reply by odbuser2** [reply]
Not a silly question.  There are a few issues with using encrypted folders:

1. Each OS supports a different means to support encrypted folders
2. The OS configuration for the encryption is usually minimal or non-existent
3. The OS does not keep with standard best practices (at least until each major update)
4. 3rd party support for encrypted folders have similar issues (unless we write it ourselves)
5. Encrypted folders are susceptible to man-in-the middle exposure (e.g. layered file systems where a layer before the encryption layer is compromised)
6. Key and salt retrieval is not customizable and often does not meet standards

But what could be done (and probably should be done on pretty much all java apps) is to use a pluggable FileSystem for all arcadedb operations (can't simply override the default one since there would be wide implications for that).  That way, any number of filesystems could be implemented (e.g. in memory, encrypted, etc.).

RandomAccessFile would also need to be pluggable and there may be limitations imposed.

Would you be willing to make the FileSystem and RandomAccessFile pluggable?

**Reply by lvca** [reply]
@odbuser2 Having a pluggable file system is certainly possible, there are only a few components that go IO with the file system. On the other side, the easiest thing would be compressing the records before they are written into the page and uncompressing them just after they are read from the file system. @odbuser2 if that is the case, where would you store the encryption credentials?

**Comment by odbuser2** [comment]
>  if that is the case, where would you store the encryption credentials?

That's a secret ;0).

Are you suggesting to encrypt while writing to the page?  I guess that would work.  If it's pluggable, the secret(s) can be stored in a number of places but I'd suggest not even addressing that.  Making a pluggable api for writing to a page made be enough and the implementor can control the entire encryption process.  But I reserve judgement until I understand what it means to write to the page.

**Reply by lvca** [reply]
ArcadeDB already supports custom event listeners for Create, Update, and Delete operations. It was missing the Read to provide a complete API to manipulate the record on read. Now it's here: https://github.com/ArcadeData/arcadedb/issues/1175. 

Example of registering an event (taken from our test suite https://github.com/ArcadeData/arcadedb/blob/3969df8c007d4b48e3e8c00385a1a4933bc6b663/engine/src/test/java/com/arcadedb/event/DatabaseEventsTest.java#L74):

```java
@Test
  public void testAfterCreate() {
    final AtomicInteger counter = new AtomicInteger();
    final AfterRecordCreateListener listener = record -> counter.incrementAndGet();

    database.getEvents().registerListener(listener);
    try {
      // CREATE SOME RECORDS HERE
    } finally {
      database.getEvents().unregisterListener(listener);
    }
  }
```

Events can be listened to at the database or type level. 

@odbuser2 you could hook up here to provide your own encryption.

**Reply by odbuser2** [reply]
This is not encryption at rest.  This is field level encryption.  The schema is not encrypted.  A pluggable file system and pluggable RandomFileAccess should allow an implementation of encryption at rest.

**Reply by lvca** [reply]
You're right, dunno why I said at rest. I renamed the test to RecordEncryptionTest.

**Comment by lvca** [comment]
With the last changes, you can implement your own Encryption at rest. Here is a full example:

https://github.com/ArcadeData/arcadedb/blob/afc283453258eec989cfc0909adef68aeb4f1d30/engine/src/test/java/com/arcadedb/event/EncryptionAtRestTest.java#L51

**Reply by dijef** [reply]
Thanks for this @lvca, it is useful. I have some observations and suggestions (I use Java TinkerPop):
1) When I read record in transaction but access property outside of transaction, it performs lazy loading: read record and fire read event. However this example uses MutableVertex and it throws outside of transaction error. It complicates the use, as not every read is modification:
`Exception in thread "" com.arcadedb.exception.TransactionException: Transaction not active
	at com.arcadedb.database.TransactionContext.getPageToModify(TransactionContext.java:313)
	at com.arcadedb.graph.ImmutableVertex.modify(ImmutableVertex.java:78)
	at org.lhasalimited.arcadedb.engine.ArcadeDbEncryption.onAfterRead(ArcadeDbEncryption.java:79)
	at com.arcadedb.database.RecordEventsRegistry.onAfterRead(RecordEventsRegistry.java:237)
	at com.arcadedb.database.LocalDatabase.invokeAfterReadEvents(LocalDatabase.java:1606)
	at com.arcadedb.database.ImmutableDocument.checkForLazyLoading(ImmutableDocument.java:182)
	at com.arcadedb.graph.ImmutableVertex.checkForLazyLoading(ImmutableVertex.java:209)
	at com.arcadedb.database.ImmutableDocument.get(ImmutableDocument.java:57)
	at com.arcadedb.gremlin.ArcadeVertex.property(ArcadeVertex.java:149)`
2) As I have to decrypt on each read and encrypt on each write for all of the fields (or selection of them), I am performing modifications that are likely to not be part of my logic

Perhaps modifying or opening `BinarySerializer` in `serializeValue` and `deserializeValue` be better option to use provided encryption/decryption logic.

**Reply by lvca** [reply]
Reading records outside a transaction is allowed, but in your case, you're modifying the record by deserializing each property. Modifying a record requires an active transaction. You could use a `record.detach()` which returns a `DetachedDocument`, namely a read-only copy of the record with all properties serialized.

Perhaps by adding encryption (same for compression), we could add a special listener into `BinarySerializer .serializeProperties` and `BinarySerializer .deserializeProperties` to convert the binary content into something else before being processed.

**Reply by dijef** [reply]
Yes, record is modified so code executing it would see decrypted values. DetachedVertex usage comes with other issues, like empty edges. My vote is for changes in serialiser then please.

**Reply by dijef** [reply]
I tried implementing encryption using events as in example you provided but it is not possible to make it work with transactions.
It's fine when reading vertex but not when writing to it. This is because encrypted value is stored inside vertex following write and same vertex is returned (it does not trigger read event as already read). 
So I added `onAfterUpdate` to change back to decrypted value, however whole transaction is written later, meaning it writes unencrypted version. Can provide full JUnit code if needed.

**Reply by dijef** [reply]
Please check test file I created.
I added boolean `USE_EXTRA_AFTER_LISTENERS` to show issue I mentioned above when set to true.
[EncryptionTest.zip](https://github.com/user-attachments/files/15911542/EncryptionTest.zip)
I think example given can't work, so changes at serialisation would be much more reliable. Please let me know if you can do them, otherwise I'll try to contribute that (I need to get work approval first).

**Reply by lvca** [reply]
@dijef I can see how encryption (and compression) would be implemented much better into the serializer. I was thinking of providing a simple listener interface that acts right at the beginning for deserialization, converting the encrypted value into a normal buffer and at the end for serialization: when the buffer is created, before returning is encrypted.

Providing a listener could be the quick solution. The best would be providing the listener and a pluggable implementation (an implementation of the listener interface) that does the job of accepting the algorithm to use, keys, etc. So everybody can just configure and use it.

The next step could be allowing to encrypt/decrypt only specific buckets (by configuration) or even only specific properties (probably overkill).

@dijef if you can find time to draft an implementation it would be awesome.

**Reply by pawellhasa** [reply]
Hello @lvca , my company approved me the contribution. Please assign to this account then. I'll share how I done it.

**Reply by lvca** [reply]
Great! Just created a new issue from this: https://github.com/ArcadeData/arcadedb/issues/1659
